types/index.d.ts

```diff
@@ -0,0 +1,166 @@
+// Minimum TypeScript Version: 4.2
+
+// NOTE: `dtslint` needs 4.2 even though these definitions should work with 4.1.
+
+declare module "hyperapp" {
+  // `app()` initiates a Hyperapp application. `app()` along with runners and
+  // subscribers are the only places where side effects are allowed.
+  function app<S>(props: App<S>): void
+
+  // `h()` builds a virtual DOM node.
+  function h<S, T extends string = string>(
+    // Tags cannot be empty strings.
+    tag: T extends "" ? never : T,
+    props: PropList<S>,
+    children?: VNode<S> | readonly VNode<S>[]
+  ): VDOM<S>
+
+  // `memo()` stores a view along with data for it.
+  function memo<
+    S,
+    D extends string | any[] | Record<string, any>
+  >(
+    view: View<D>,
+    data: D
+  ): VDOM<S>
+
+  // `text()` creates a virtual DOM node representing plain text.
+  function text<T, S>(
+    value: T extends (symbol | ((..._: any[]) => any)) ? never : T,
+    node?: Node
+  ): VDOM<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A Hyperapp application instance has an initial state and a base view.
+  // It must also be mounted over an available DOM element.
+  type App<S> = Readonly<{
+    init: StateFormat<S> | Action<S>
+    view: View<S>
+    node: Node
+    subscriptions?: Subscriptions<S>
+    middleware?: Middleware<S>
+  }>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View<S> = (state: State<S>) => VDOM<S>
+
+  // The subscriptions function manages a set of subscriptions.
+  type Subscriptions<S> = (state: State<S>) => Subscription<S>[]
+
+  // A subscription represents subscriber activity.
+  type Subscription<S, D = any> = boolean | undefined | SubscriberDescriptor<S, D> | Unsubscribe
+
+  // A subscriber reacts to subscription updates.
+  type SubscriberDescriptor<S, D> = [Subscriber<S, D>, Payload<D>]
+  type Subscriber<S, D> = (dispatch: Dispatch<S>, props?: Payload<D>) => void | Unsubscribe
+
+  // An unsubscribe function cleans up a canceled subscription.
+  type Unsubscribe = () => void
+
+  // Middleware allows for custom processing during dispatching.
+  type Middleware<S> = (dispatch: Dispatch<S>) => Dispatch<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A dispatched action handles an event in the context of the current state.
+  type Dispatch<S> = (action: Action<S>, props?: Payload<any>) => void
+
+  // An action transforms existing state and/or wraps another action.
+  type Action<S, P = any> = ActionTransform<S, P> | ActionDescriptor<S, P>
```

Comment:
Incorrect, considering later down you bind all `onsomething` events to `Action<S, WhateverPayloadFromEvent>`

This definition of action means: `onclick: [(state, x) => ({...state, foo: state.foo + x}), 42]`  would be invalid when it is clearly not. 

I would suggest changing it to `Action<S, P> = ActionTransform<S,P>| ActionDescriptior<S>`  and then use a validator function to ensure that payloads given in action descriptors match the required type for the ActionTransform. (Like I did in my set of types)

Reply:
Could you expand your `onclick` example into a test case ?

Reply:
I tried, but I got stuck on not being able to define an action. Gonna see if I can work around it

Reply:
Ok here's the test:

```ts
import { app, h, text, ActionTransform } from "hyperapp"

type State = { foo: number }

const AddSome: ActionTransform<State, number> = (state, amount) => ({
    ...state,
    foo: state.foo + (amount || 0),
})

//$ExpectError
h("button", { onclick: AddSome }, text("add event?!"))

//$ExpectType VNode
h("button", { onclick: [AddSome, 32] }, text("add 32"))

//$ExpectError
h("button", { onclick: [AddSome, "foo"] }, text("add string?!"))
```

Reply:
The part that fails now is the middle button, because `AddSome` does not take `MouseEvent` for a payload.

Reply:
This custom payload issue is a great catch !

Reply:
I've yet to solve that third case with `[AddSome, "foo"]`.  Were you able to solve that in your types ?

Reply:
Addressed in https://github.com/jorgebucaran/hyperapp/pull/1047.